Add support for NERSC CFS readonly mount

sbailey opened the pull request:
This PR implements #1842 by adding support for $DESI_ROOT_READONLY to specify the read-only /dvs_ro/cfs/cdirs/desi mount of /global/cfs/cdirs/desi.
* new `findfile(..., readonly=False/True)` option to request the read-only filepath
* new function `desispec.io.meta.get_readonly_filepath(filepath)` to convert a read-write filepath into a read-only filepath
* new helper function `desispec.io.meta.get_desi_root_readonly()` used by `findfile` and `get_readonly_filepath` that checks for $DESI_ROOT_READONLY, checks that it actually exists, and caches the answer so that subsequent calls to `findfile` and `get_readonly_filepath` don't have to keep re-checking.

In all cases, if $DESI_ROOT_READONLY isn't set or is set to a value that doesn't exist (e.g. /dvs_ro/cfs/cdirs is mounted on cori compute nodes but not login nodes), the functions default to providing the read-write path.  Usage of this is opt-in, i.e. the default is `findfile(..., readonly=False)`, returning the regular read-write path even if $DESI_ROOT_READONLY is set.

I updated the desi_proc, desi_proc_joint_fit, and desi_proc_tilenight scripts to use this for input files.  I did *not* update the redshift scripts since they are bash scripts oriented towards cd'ing to a top-level directory and then working relative to that for both input and output, and would require more surgery to make use of this.

For the record: I also considered updating all `desispec.io.read_*` functions to convert input filenames to read-only versions, but I decided that could be one level too clever since then it would be reading a different filepath than you had actually requested, which smells dangerous even if the read-only path is supposed to be the same.

Improvements are modest but noticeable, e.g. read_frame calls in processing 20220914/00142434 have fewer entries in the high tail (though in this run the max time was actually from the read-only mount, not the read-write mount):

![image](https://user-images.githubusercontent.com/218471/191850532-898aac72-0abe-4230-85a2-0d24f162685a.png)

This hopefully will buy us some additional robustness against NERSC I/O instabilities, though it isn't a silver bullet.

I've tested this on individual runs of desi_proc and desi_proc_tilenight, and am now doing a run of a partial night to include ccdcalib, arc, psfnight, flat, nightlyflat, tilenight science, and redshifts in /global/cfs/cdirs/desi/users/sjbailey/spectro/redux/readonly/ for night 20220515 (all cals but subset of tiles).  I'm still finding and catching typos/bugs there, but I think this is far enough enough for code review.

@akremin please review this from a desi pipeline perspective.

@dmargala please review this from a NERSC I/O usage perspective (also see NERSC ticket  INC0191433 for related questions, pending answers from Lisa).

akremin reviewed (approved):
The changes look good, but there are enough of them that I'll let the tests your running determine when this has converged. The basic mechanisms look good and it's a matter of finding all caveats of filenames that actually need write access and fail. 

My three comments don't necessarily require changes, so I will go ahead and approve this now. Merge whenever the end-to-end tests succeed.

sbailey commented on py/desispec/io/meta.py:
```diff
@@ -17,14 +17,65 @@
+    elif 'DESI_ROOT_READONLY' in os.environ:
+        if os.path.exists(os.environ['DESI_ROOT_READONLY']):
+            _desi_root_readonly = os.environ['DESI_ROOT_READONLY']
+            log.debug("Using $DESI_ROOT_READONLY=%s", _desi_root_readonly)
```
Formatting+argument will only evaluate the string inside the logger call if the log level is high enough that it knows it will will print the string, while f-string formatting will always happen before being passed into the logger call.  Thus format+argument(s) is preferred for debug-level calls that normally aren't printed.  We've had cases in the past were significant time was spent on f-string formatting of debug messages that never got printed deep within loops, and switching to format+arguments was faster.  Some people strictly use format+args for any log messages for this reason; I tend to to it only for debug messages because I overall prefer the clarity of f-string formatting and performance-wise it doesn't matter if the message is going to be printed either way.

akremin replied:
Thank you for the explanation, that makes perfect sense.

sbailey replied:
as a paranoia check I also tried running the unit tests with DESI_LOGLEVEL=debug to ensure that I didn't have a typo in my debug log statements...

sbailey commented on py/desispec/scripts/proc.py:
```diff
@@ -728,7 +730,7 @@ def main(args=None, comm=None):
                         desispec.scripts.specex.mean_psf(psfs,psfnightfile)
                 if os.path.isfile(psfnightfile) : # now use this one
-                    input_psf[camera] = psfnightfile
+                    input_psf[camera] = get_readonly_filepath(psfnightfile)
```
Good point, though when I went to implement it I noticed that this whole section is wrapped in an `if False:` block left over from when desi_proc was trying to do multi-exposure tasks that eventually got replaced with desi_proc_joint_fit.  It's not strictly related to this PR but I'll remove that unused block here rather than craft read-only logic for code that is never run.

sbailey commented on py/desispec/io/meta.py:
```diff
@@ -17,14 +17,65 @@
+    log = get_logger()
+    if _desi_root_readonly is not None:
+        log.debug('Using cached _desi_root_readonly=%s', _desi_root_readonly)
+        return _desi_root_readonly
```
You are correct that line 33 isn't strictly needed given the current structure of the code, but I put it here for robustness to future updates that might change the elif/else or post-else logic and potentially break this case.  i.e. I'm going to leave this one as-is.

sbailey commented:
For the record: NERSC consulting confirmed that the /dvs_ro caching is based upon inode not filename, so we are safe for the case of deleting old files and re-running to generate new files with the same name but different inode.  i.e. that won't accidentally pick up a cached value of a deleted file.

I'll repost when the end-to-end test is finished and vetted.  There were a number of job hiccups that led to additional code changes for further robustness to I/O problems.

sbailey commented:
End-to-end tests in /global/cfs/cdirs/desi/users/sjbailey/spectro/redux/readonly have finished, testing ccdcalib, arc, psfnight, flat, nightlyflat, tilenight, and redshifts.  Documenting a few changes since @akremin's review:

* Fixed one readonly-related bug for a PSF file -- a combo of `findfile` + `inpsf.replace(...)` were resulting in a readonly path to a file that needed to be renamed, now fixed.
* Use tempfiles for preproc and extraction bundle files, to avoid leaving files with the correct name but incorrect contents in the case of a crash/timeout during write.
* Removed a block in `proc.py` for multi-exposure PSF fitting, which had been "commented out" with `if False:` (i.e. rather than crafting readonly file usage for unused code, I removed it).
* Added try/except around extractions so that a failure doesn't cause the job to hang until timeout.

sbailey commented:
verbally confirmed signoff with @akremin ; merging now.
